server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,911 @@
+			.filter(item => categorized[item].length > 0)
+			.map(item => `${STATUS_TITLES[item]} (${categorized[item].length})`);
+
+		let buf = `<h3>Your friends: <small> `;
+		if (sorted.length > 0) {
+			buf += `Total (${friends.size}) | ${sorted.join(' | ')}`;
+		} else {
+			buf += `</h3><em>you have no friends added on Showdown lol</em><br /><br /><br />`;
+			buf += `<strong>To add a friend, use </strong><code>/friend add [username]</code>.<br /><br />`;
+			buf += `<strong>To move over your friends to this account from a different account, `;
+			buf += `sign into that account and use </strong><code>/friend requesttransfer [new name]</code>.`;
+			return buf;
+		}
+		buf += `</h3> `;
+
+		for (const key in categorized) {
+			const friendArray = categorized[key].sort();
+			if (friendArray.length === 0) continue;
+			buf += `<h4>${STATUS_TITLES[key]} (${friendArray.length})</h4>`;
+			for (const friend of friendArray) {
+				buf += `<div class="pad"><div>`;
+				buf += this.displayFriend(friend, loginTimes[toID(friend)]);
+				buf += `</div></div>`;
+			}
+		}
+
+		return buf;
+	}
+	displayFriend(userid: string, login?: number) {
+		const user = Users.getExact(userid); // we want this to be exact
+		const connected = user?.connected;
+		const name = Utils.escapeHTML(user ? user.name : userid);
+		const statusType = connected ?
+			`<strong style="color:${STATUS_COLORS[user!.statusType]}">\u25C9 ${STATUS_TITLES[user!.statusType]}</strong>` :
+			'\u25CC Offline';
+		let buf = connected ?
+			`<span class="username"> <strong>${name}</strong></span><span><small> (${statusType})</small></span>` :
+			Utils.html`<i>${name}</i> <small>(${statusType})</small>`;
+		buf += `<br />`;
+
+		const oldName = this.renames.get(userid);
+		if (oldName) {
+			buf += Utils.html`<small>(recently renamed from ${oldName})</small><br />`;
+		}
+
+		const curUser = Users.get(userid); // might be an alt
+		if (user?.connected) {
+			if (user.userMessage) buf += Utils.html`Status: <i>${user.userMessage}</i><br />`;
+		} else if (curUser?.id && curUser.id !== userid) {
+			buf += `<small>On an alternate account</small><br />`;
+		}
+		if (login && !connected) {
+			// THIS IS A TERRIBLE HACK BUT IT WORKS OKAY
+			const time = Chat.toTimestamp(new Date(Number(login)), {human: true});
+			buf += `Last login: ${time.split(' ').reverse().join(', on ')}`;
+			buf += ` (${Chat.toDurationString(Date.now() - login, {precision: 1})} ago)`;
+		}
+		buf = `<div class="infobox">${buf}</div>`;
+		return toLink(buf);
+	}
+	async addFriend(senderName: User | string, receiverName: string) {
+		const receiverID = toID(receiverName);
+		const senderID = toID(senderName);
+		if (!receiverID) throw new Chat.ErrorMessage(`Invalid user.`);
+		return this.query({
+			type: 'transaction', statement: 'add', data: [receiverID, senderID],
+		});
+	}
+	async removeFriend(userName: string, friendName: string) {
+		const userid = toID(userName);
+		const friendID = toID(friendName);
```

Comment:
I think I left a similar comment on this PR before, but maybe I didn't properly explain.

Internally, we work entirely with IDs, right? The only time you can encounter something that isn't an ID is when interfacing with the user. So do `toID` once, when parsing human input (i.e. only @ in the chat commands). Any library functions should take `ID` as their type instead of `string`, and not redundantly call `toID` on their inputs.